job/replay.go

```diff
@@ -0,0 +1,187 @@
+package job
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/odpf/optimus/core/cron"
+	"github.com/odpf/optimus/core/multi_root_tree"
+	"github.com/odpf/optimus/models"
+	"github.com/pkg/errors"
+)
+
+const (
+	// DateFormat YYYY-mm-dd for replay dates and dag start date
+	DateFormat = "2006-01-02"
+)
+
+func (srv *Service) Replay(namespace models.NamespaceSpec, replayJobSpec models.JobSpec, dryRun bool, start, end time.Time) (*models.DAGNode, error) {
+	projectJobSpecRepo := srv.projectJobSpecRepoFactory.New(namespace.ProjectSpec)
+	jobSpecs, err := srv.getDependencyResolvedSpecs(namespace.ProjectSpec, projectJobSpecRepo, nil)
+	if err != nil {
+		return nil, err
+	}
+	dagSpecMap := make(map[string]models.JobSpec)
+	for _, currSpec := range jobSpecs {
+		dagSpecMap[currSpec.Name] = currSpec
+	}
+
+	rootInstance, err := prepareTree(dagSpecMap, replayJobSpec.Name, start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	if dryRun {
+		//if only dry run, exit now
+		return rootInstance, err
+	}
+
+	return rootInstance, nil
+}
+
+// prepareTree creates a execution tree for replay operation
+func prepareTree(dagSpecMap map[string]models.JobSpec, replayJobName string, start, end time.Time) (*models.DAGNode, error) {
+	replayJobSpec, found := dagSpecMap[replayJobName]
+	if !found {
+		return nil, fmt.Errorf("couldn't find any job with name %s", replayJobName)
+	}
+
+	// compute runs that require replay
+	tree := multi_root_tree.NewMultiRootDAGTree()
+	parentNode := models.NewDAGNode(replayJobSpec)
+	if runs, err := getRunsBetweenDates(start, end, replayJobSpec.Schedule.Interval); err == nil {
+		for _, run := range runs {
+			parentNode.Runs.Add(run)
+		}
+	} else {
+		return nil, err
+	}
+	tree.AddNode(parentNode)
+
+	rootInstance, err := populateDownstreamDAGs(tree, replayJobSpec, dagSpecMap)
+	if err != nil {
+		return nil, err
+	}
+
+	rootInstance, err = populateDownstreamRuns(rootInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	return rootInstance, nil
+}
+
+func findOrCreateDAGNode(tree *multi_root_tree.MultiRootDAGTree, dagSpec models.JobSpec) *models.DAGNode {
+	node, ok := tree.GetNodeByName(dagSpec.Name)
+	if !ok {
+		node = models.NewDAGNode(dagSpec)
+		tree.AddNode(node)
+	}
+	return node
+}
+
+func populateDownstreamDAGs(tree *multi_root_tree.MultiRootDAGTree, jobSpec models.JobSpec, dagSpecMap map[string]models.JobSpec) (*models.DAGNode, error) {
+	for _, childSpec := range dagSpecMap { // (1) -> 2 -> 3
+		childNode := findOrCreateDAGNode(tree, childSpec) // 3
+		for _, depDAG := range childSpec.Dependencies {
+			var isExternal = false
+			parentSpec, ok := dagSpecMap[depDAG.Job.Name] // 2
+			if !ok {
+				if depDAG.Type == models.JobSpecDependencyTypeIntra {
+					return nil, errors.Wrap(ErrJobSpecNotFound, depDAG.Job.Name)
+				} else {
```

Comment:
we can pull the else statement out if we are returning in the base if statement.

Reply:
yes, removed